tools/Modules/TestFx-Tasks.psm1

```diff
@@ -0,0 +1,436 @@
+    }
+
+    $azSpn = Get-SPN -ADAppId $azAdApp.ApplicationId
+
+    if(($azAdApp -ne $null) -and ($azSpn -eq $null))
+    {
+        if($createADSpnIfNotFound -eq $true)
+        {
+            $azSpn = New-AzureRmADServicePrincipal -ApplicationId $azAdApp.ApplicationId            
+            Write-Host "Waiting for SPN to fully create and able to query ....."
+
+            for($i=0; $i -lt 3; $i++)
+            {
+                #need to find a better deterministic way
+                #Cases are when you try to assign roles to newly created SPN, it is not able to find it and hence Role Assignment fails
+                #Get-SPN -ADAppId $azAdApp.ApplicationId
+                Start-Sleep -Seconds 5
+            }
+        }
+    }
+
+    if($azSpn -ne $null)
+    {
+        $spnScope = [string]::Format("{0}{1}", "/subscriptions/", "$SubscriptionId/")
+
+        $rdStr = [string]::Format("{0} Role will be assigned to SPNName:{1} for scope {2}", "Reader", $azAdApp.ApplicationId, $spnScope)
+        Write-Host $rdStr
+
+        $roleDef = Get-AzureRmRoleAssignment -ServicePrincipalName $azAdApp.ApplicationId -Scope $spnScope
+        if($roleDef -eq $null)
+        {
+            New-AzureRmRoleAssignment -RoleDefinitionName "Reader" -ServicePrincipalName $azAdApp.ApplicationId -scope $spnScope
+        }
+    }
+
+    Get-ServicePrincipalDetails -AdApp $azAdApp -AdSpn $azSpn
+
+    Write-Host "Log onto Azure Portal and use Above SPN Details to find your Service Principal and obtain SPN Secret/ Authentication Key/ Secret Key for the newly created Service Principal" -ForegroundColor DarkYellow
+    Write-Host "Visit https://docs.microsoft.com/en-us/azure/resource-group-create-service-principal-portal" -ForegroundColor DarkYellow
+}
+
+Function Get-ADApp
+{
+    param(
+    [Parameter(Mandatory=$true, HelpMessage="Azure AD Apllication Display Name")]
+    [ValidateNotNullOrEmpty()]
+    [string]$ADAppDisplayName
+    )
+
+    $displayName = $ADAppDisplayName
+    $homePage = "http://www.$displayName.psmodule"
+    $identityUri = "http://$displayName"
+
+    $azAdApp = Get-AzureRmADApplication -IdentifierUri $identityUri
+
+    $PsCmdLet.WriteObject($azAdApp)
+}
+
+[CmdletBinding]
+Function Get-SPN
+{
+    param(
+    [Parameter(Mandatory=$true, ParameterSetName="DisplayName", HelpMessage="Azure AD Apllication Display Name")]
+    [ValidateNotNullOrEmpty()]
+    [string]$ADAppDisplayName,
+
+    [Parameter(Mandatory=$true, ParameterSetName="AppId", HelpMessage="Azure AD Apllication ID")]
+    [ValidateNotNullOrEmpty()]
+    [string]$ADAppId
+    )
+
+    $appId = ""
+    if([string]::IsNullOrEmpty($ADAppDisplayName) -eq $false)
+    {
+        $azAdApp = Get-ADApp -ADAppDisplayName $ADAppDisplayName
+        $appId = $azAdApp.ApplicationId
+    }
+    
+    if([string]::IsNullOrEmpty($ADAppDisplayName) -eq $true)
+    {
+        $appId = $ADAppId
+    }
+
+    $azSpn = Get-AzureRmADServicePrincipal -ServicePrincipalName $appId
```

Comment:
Normally you have to create a service principal after creating the appi, see: https://gist.github.com/devigned/dae74a7ca54000f7b714#file-sp-w-cert-azps-1-0-ps1-L30